base/compiler/ssair/passes.jl

```diff
@@ -649,64 +695,76 @@ function perform_lifting!(compact::IncrementalCompact,
         lf.need_argupdate || continue
-        for i = 1:length(old_node.edges)
-            edge = old_node.edges[i]
-            isassigned(old_node.values, i) || continue
-            val = old_node.values[i]
-            if is_old(compact, old_node_ssa) && isa(val, SSAValue)
-                val = OldSSAValue(val.id)
-            end
-            if isa(val, AnySSAValue)
-                val = simple_walk(compact, val)
-            end
-            if val in keys(lifted_leaves)
-                push!(new_node.edges, edge)
-                lifted_val = lifted_leaves[val]
-                if lifted_val === nothing
+
+        if isa(lf.node, PhiNode)
+            old_node = compact[old_node_ssa][:inst]::PhiNode
+            new_node = lf.node::PhiNode
+            should_count = !isa(lf.ssa, OldSSAValue) || already_inserted(compact, lf.ssa)
+            for i = 1:length(old_node.values)
+                isassigned(old_node.values, i) || continue
+                val = lifted_value(compact, old_node_ssa, old_node.values[i],
+                                   lifted_philikes, lifted_leaves, reverse_mapping)
+                val !== SKIP_TOKEN && push!(new_node.edges, old_node.edges[i])
+                if val === UNDEF_TOKEN
                     resize!(new_node.values, length(new_node.values)+1)
-                    continue
+                elseif val !== SKIP_TOKEN
+                    should_count && _count_added_node!(compact, val)
+                    push!(new_node.values, val)
                 end
-                val = lifted_val.val
-                if isa(val, AnySSAValue)
-                    callback = (@nospecialize(pi), @nospecialize(idx)) -> true
-                    val = simple_walk(compact, val, callback)
-                end
-                should_count && _count_added_node!(compact, val)
-                push!(new_node.values, val)
-            elseif isa(val, AnySSAValue) && val in keys(reverse_mapping)
-                push!(new_node.edges, edge)
-                newval = lifted_phis[reverse_mapping[val]].ssa
-                should_count && _count_added_node!(compact, newval)
-                push!(new_node.values, newval)
-            else
-                # Probably ignored by path condition, skip this
             end
+        elseif isa(lf.node, IfElseCall)
+            ifelse_call = lf.node.call::Expr
+            then_result, else_result = ifelse_call.args[3], ifelse_call.args[4]
+            then_result = lifted_value(compact, old_node_ssa, then_result,
+                                       lifted_philikes, lifted_leaves, reverse_mapping)
+            else_result = lifted_value(compact, old_node_ssa, else_result,
+                                       lifted_philikes, lifted_leaves, reverse_mapping)
+            @assert then_result !== SKIP_TOKEN && then_result !== UNDEF_TOKEN
+            @assert else_result !== SKIP_TOKEN && else_result !== UNDEF_TOKEN
```

Comment:
I haven't convinced myself yet that these asserts will always be true

Reply:
As far as I understand, we should always have `then_result` and `else_result` within the keys of `lifted_leaves`, so `lifted_value` should never return `SKIP_TOKEN` here. `UNDEF_TOKEN` is only used for `:new` expression, so it does not matter.